Protocols: Create posix.Symlink to create links on download; Fix #5319

snafus opened the pull request:
Subclass of the Default POXIS implementation to allow a symlink to be made to the remote location
when issuing at 'get' of the remote file.

Only the get() method is overridden to provide this functionality. Other methods use the default implementation,
which involve using the remote file directly.

mlassnig commented on lib/rucio/rse/protocols/posix.py:
```diff
@@ -16,7 +16,9 @@
 import os
 import os.path
 import shutil
+import logging
```
please sort imports alphabetically

mlassnig reviewed (commented):
tiny comment. and you might want to add a unittest?

snafus committed:
Protocols: Create posix.Symlink to create links on download rucio #5319

Subclass of the Default posix implemenation to allow a symlink to be made to the remote location,
instead of copying the remote file locally.

Only the get() method is overridden to provide this functionality. Other methods use the default implementation,
which involve using the remote file directly.

snafus commented:
Hi @mlassnig; Sorry for the delay, and the problems above. The code was refactored a little to allow (python 2) tests to pass and retested. James

mlassnig commented on lib/rucio/rse/protocols/posix.py:
```diff
@@ -210,3 +211,46 @@ def stat(self, pfn):
+            self.logger(logging.ERROR, 'Symlink {} appears to be a broken link to {}'
+                        .format(dest, path))
+            if os.lstat(dest) and os.path.islink(dest):
+                call(['rm', '-rf', dest])
```
this can't be done with shutil.rmtree() ?

calling rm -rf seems so ... :scream:

snafus replied:
I was attempting consistency with (in the Default implementation): https://github.com/rucio/rucio/blob/2c9d77f203fbe6e6ec78b415ccfe457d206b0519/lib/rucio/rse/protocols/posix.py#L77 

shutil.rmtree() does not work with symlinks (afaik).  
os.unlink(dest) might be the simplest option here, (and I've put in this commit)? 

I could replace the 'rm' in the Default class implementation with an shutil.rmtree, if you wanted here?

mlassnig commented on lib/rucio/rse/protocols/posix.py:
```diff
@@ -210,3 +211,46 @@ def stat(self, pfn):
+                        'Symlink {} created for {} from {}'
+                        .format(dest, path, pfn))
+        except Exception as e:
+            print("JW,", e)
```
remove?

snafus replied:
Argh - thanks. I removed the print statement (unless you were suggesting the general exception?).

snafus replied:
Removed also the generic exception, which should propagate up higher.
